server/text_generation_server/models/ct2_causal_lm.py

```diff
@@ -0,0 +1,405 @@
+# coding=utf-8
+# Copyright 2023 Michael Feil. All rights reserved.
+#
+# This code is loosely based on Huggingface text-generation-inference v0.9.3's causal_lm.py implementation.
+# While it remains licensed under Apache License, Version 2.0,
+# text-generation-inference itself on 7/28/2023 has changed its license.
+# This code remains unaffected by this, and contributing it to a 
+# upstream repo with new License was not intentional.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import numpy as np
+import os
+import multiprocessing
+from pathlib import Path
+
+from huggingface_hub.constants import HUGGINGFACE_HUB_CACHE
+from opentelemetry import trace
+from transformers import (
+    AutoTokenizer,
+    AutoConfig,
+)
+from typing import Optional, Tuple, List, Type, Dict
+
+from text_generation_server.models import Model
+from text_generation_server.models.types import (
+    PrefillTokens,
+    Generation,
+    GeneratedText,
+)
+
+from text_generation_server.utils import Sampling
+from text_generation_server.models.causal_lm import CausalLMBatch
+
+try:
+    import ctranslate2
+except ImportError:
+    ctranslate2 = None
+
+
+tracer = trace.get_tracer(__name__)
+
+
+class CT2CausalLM(Model):
```

Comment:
I am a little worried about the diverge between `CT2CausalLM` and `CausalLM`. I would if we could reuse code between `CT2CausalLM` and `CausalLM` as much as possible.

Could review https://github.com/Preemo-Inc/text-generation-inference/pull/5 ? I think it could be the first step to reuse code between models.

Reply:
Can be done - also better in a PR that refactors all models. The API of `forward` is not compatible, as CT2 uses different input tensor (all past input IDs) compared to the transformers implentation.